### DISCUSSION TITLE
Minimum Requirements for dev

### DISCUSSION THREAD

**Discussion by danlangford** [pappde/bmai#37] [General]
What should be the minimum version of various tools that we use on the build stack? 

There have been times in the past that we have set min version of cmake to 3.8 to support Visual Studio 2017. VS2017 supports up to CMake 3.12. The CMake config is set up for c++14. 

@pappde Are you still on VS2017? should we remain on c++14? 

how do you feel about enforcing a code style? though it can be difficult to adapt if you can enforce a certain code style you can later on know that all diffs are logic only and never just somebodies development preferences.

**Comment by pappde** [comment]
@danlangford  I use VS2022. We can definitely increase the requirement, though not sure what we gain from it.

Code Style: there is a code style already in place in the codebase which should be followed, as with any codebase. We can certainly document and enforce. Or are you suggesting adopting a different convention?

**Reply by danlangford** [reply]
is the current code style a pre-existing standard that my IDE can help me enforce?
here are the ones that come baked into CLion
![image](https://github.com/pappde/bmai/assets/721364/0f90db54-39db-470f-b5aa-e4af361e3318)
but others can be loaded
if we can encode the standard then I can get help following it and even (if we want) fail a build if any source isn't following it.

**Reply by danlangford** [reply]
I think it would be good to at least encode your preferred style into a `.clang-format` file. if your IDE is set up in your preferred manner (as it relates to tabs vs spaces, brace placement, etc.) will you export your settings and send me over the `.vssettings` file? 

> - Access the Import and Export Settings Wizard:  
Go to the menu bar and select Tools > Import and Export Settings....
> - Choose Export:  
In the wizard, select Export selected environment settings and click Next.
> - Select Settings to Export:  
Choose the settings you want to export. You can select all settings or specific ones like Text Editor settings which include tab size and bracket placement.

**Reply by pappde** [reply]
This is an old project, but here are the basics for the naming style it uses:

- PascalCase for classes
- PascalCase for methods
- lowercase_underscore for variables and parameters
- m_lowercase_underscore for properties
- UPPERCASE_UNDERSCORE for constants/enums
- no namespaces, instead classes have a prefix (e.g. "BMC_")

Braces: no braces for a single line block, otherwise braces on new lines. 

I'm not sure if this matches an official style. Copilot says this is close to Google, but that uses m_camelCase for properties.

**Comment by danlangford** [comment]
@pappde 
> VS2022. We can definitely increase the requirement, though not sure what we gain from it.

well CMake gets a boost from 3.12 to at least 3.26. there are a lot of new features and bugs fixed from 3.12-3.26.

while VS2017 did get ability to support c++17 eventually, there is much more solid support (from what I understand) in vs2019 and vs2022. the filesystem namespace is a huge win and overall the language is evolving to be more compact, expressive, and easy to read.

I also have been finding modern solutions to some of my questions and have been needing to dig deeper for more C or C++11/14 solutions which isn't always easy (back to the missing `filesystem`)

ill look at bumping up some version numbers and we can see how things go
thanks

**Comment by pappde** [comment]
If it makes it easier to maintain, then it's worth seeing if updating the C++ version is easy. 

I did briefly look at changing the C++ version to 20 and spotted a couple of things that needed to be fixed. Added commit aa39a0a

At this point, I believe you can safely change it to use a later C++ version.
